contrib/tester-progs/fork-tester.c

```diff
@@ -2,99 +2,80 @@
 
 #include <stdio.h>
 #include <stdlib.h>
-#include <unistd.h>
-#include <sys/types.h>
-#include <sys/socket.h>
+#include <sys/prctl.h>
 #include <sys/wait.h>
-#include <arpa/inet.h>
-#include <fcntl.h>
-#include <signal.h>
-
-// NB: global pipe for child 2 to notify parent that it has finished.
-int Pipe[2];
+#include <unistd.h>
 
-void client_run()
+void child2(pid_t reaper_pid)
 {
-	// just print a message
-	printf("child 2 (pid:%d, ppid:%d) starts\n", getpid(), getppid());
-	printf("child 2 done\n");
-	return;
+	pid_t initial_ppid = getppid();
+	pid_t pid = getpid();
+
+	printf("child 2 (pid:%d, ppid:%d) starts\n", pid, initial_ppid);
+
+	pid_t new_ppid;
+	for (int i = 0;; i++) {
+		new_ppid = getppid();
+		if (new_ppid == reaper_pid) {
+			break;
+		}
+		if (i == 30) {
+			fprintf(stderr, "giving up on waiting our parent to die\n");
+			exit(EXIT_FAILURE);
+		}
+		sleep(1);
+	}
+	printf("child 2 (pid:%d, ppid:%d) exits\n", pid, new_ppid);
 }
 
-void child1()
+void child1(pid_t ppid)
 {
 	pid_t pid;
 
-	if  ((pid = fork()) == -1) {
+	if ((pid = fork()) == -1) {
 		perror("fork");
 		exit(1);
 	} else if (pid == 0) {
-		for (int i=0; ;i++) {
-			int ppid = getppid();
-			if (ppid == 1) {
-				break;
-			}
-			if (i == 30) {
-				fprintf(stderr, "giving up on waiting our parent to die\n");
-				exit(1);
-			}
-			sleep(1);
-		}
-		client_run();
+		child2(ppid);
 	} else {
-		/* chilid 1 exits, after creating child 2 */
+		/* child 1 exits, after creating child 2 */
 		printf("child 1 (pid:%d) exits\n", getpid());
 		return;
 	}
-
 }
 
-void alarm_handler(int signum)
+void wait_childrens(void)
 {
-	fprintf(stderr, "got an alarm, bailing out\n");
-	exit(1);
+	while (1) {
+		int status;
+		int pid = wait(&status);
+		if (pid == -1) {
+			printf("parent (pid:%d) no more descendants\n", getpid());
+			break;
+		} else {
+			printf("parent (pid:%d) child (%d) exited with: %d\n", getpid(), pid,
+			       status);
+		}
```

Comment:
Can you add a sleep here? Having tight loops without sleeps where we print something can be problematic.

Reply:
I had no problem while testing because `wait` syscall is blocking as suggested in the documentation.
For me it is unnecessary to sleep. What kind of problem were you thinking of ?

```
# https://man7.org/linux/man-pages/man2/wait.2.html
If a child has already changed state, then these calls return immediately. 
Otherwise, they block until either a child changes state or a signal handler interrupts the call
```

Reply:
I see. Yap, that makes sense to me. Please ignore my comment above.